File: backend/api/routes/reports.py
"""报告管理路由"""

from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from backend.db import get_db
from backend.db.repository import ReportRepository
from backend.db.schemas import ReportResponse, ReportDetailResponse, StepResponse, AssertionResultResponse
from backend.core.report_service import ReportService


router = APIRouter(prefix="/reports", tags=["reports"])


def get_report_repo(db: AsyncSession = Depends(get_db)) -> ReportRepository:
    return ReportRepository(db)


@router.get("", response_model=dict)
async def list_reports(
    status: str = Query(default="all"),
    date: str = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=100),
    report_repo: ReportRepository = Depends(get_report_repo),
):
    """获取报告列表"""
    reports, total = await report_repo.list(
        status=status,
        date=date,
        page=page,
        page_size=page_size,
    )
    return {
        "reports": [ReportResponse.model_validate(r) for r in reports],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.get("/{report_id}", response_model=ReportDetailResponse)
async def get_report(
    report_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取报告详情

    支持通过 report_id 或 run_id 查询报告。
    优先使用 report_id，如果未找到则尝试使用 run_id。
    """
    report_repo = ReportRepository(db)
    report = await report_repo.get(report_id)

    # If not found by report_id, try by run_id
    if not report:
        report = await report_repo.get_by_run_id(report_id)

    if not report:
        raise HTTPException(status_code=404, detail="Report not found")

    # Use ReportService for complete data
    report_service = ReportService(db)
    data = await report_service.get_report_data(report.run_id)

    if not data:
        raise HTTPException(status_code=404, detail="Report data not found")

    # Transform steps
    step_responses = [
        StepResponse(
            id=s.id,
            run_id=s.run_id,
            step_index=s.step_index,
            action=s.action,
            reasoning=s.reasoning,
            screenshot_url=f"/api/runs/{s.run_id}/screenshots/{s.step_index}" if s.screenshot_path else None,
            status=s.status,
            error=s.error,
            duration_ms=s.duration_ms,
            created_at=s.created_at,
        )
        for s in data["steps"]
    ]

    # Transform assertion results
    def transform_assertion_results(results):
        return [
            AssertionResultResponse(
                id=ar.id,
                run_id=ar.run_id,
                assertion_id=ar.assertion_id,
                status=ar.status,
                message=ar.message,
                actual_value=ar.actual_value,
                created_at=ar.created_at,
            )
            for ar in results
        ]

    return ReportDetailResponse(
        id=report.id,
        run_id=report.run_id,
        task_id=report.task_id,
        task_name=report.task_name,
        status=report.status,
        total_steps=report.total_steps,
        success_steps=report.success_steps,
        failed_steps=report.failed_steps,
        duration_ms=report.duration_ms,
        created_at=report.created_at,
        steps=step_responses,
        assertion_results=transform_assertion_results(data["assertion_results"]),
        ui_assertion_results=transform_assertion_results(data["ui_assertion_results"]),
        pass_rate=data["pass_rate"],
        precondition_results=data.get("precondition_results"),
        timeline_items=data.get("timeline_items"),
    )

File: frontend/src/api/reports.ts
// frontend/src/api/reports.ts
import { apiClient } from './client'
import type { Report, Step, AssertionResult } from '../types'

interface StepApiResponse {
  id: string
  run_id: string
  step_index: number
  action: string
  reasoning: string
  screenshot_url: string | null
  status: string
  error: string | null
  duration_ms: number
  created_at: string
}

interface ReportApiResponse {
  id: string
  run_id: string
  task_id: string
  task_name: string
  status: string
  total_steps: number
  success_steps: number
  failed_steps: number
  duration_ms: number
  created_at: string
}

interface AssertionResultApiResponse {
  id: string
  run_id: string
  assertion_id: string
  status: string
  message: string | null
  actual_value: string | null
  created_at: string
}

interface ReportDetailApiResponse extends ReportApiResponse {
  steps: StepApiResponse[]
  assertion_results: AssertionResultApiResponse[]
  ui_assertion_results?: AssertionResultApiResponse[]
  api_assertion_results?: AssertionResultApiResponse[]
  pass_rate?: string
  api_pass_rate?: string
}

interface ReportsListApiResponse {
  reports: ReportApiResponse[]
  total: number
  page: number
  page_size: number
}

export interface ReportsListParams {
  status?: string
  date?: string
  page?: number
  page_size?: number
}

// 转换 API 响应为前端类型
function transformReport(report: ReportApiResponse): Report {
  return {
    ...report,
    status: report.status as 'success' | 'failed',
  }
}

const API_BASE_FOR_IMAGES = import.meta.env.VITE_API_BASE?.replace('/api', '') || 'http://localhost:8080'

function transformStep(step: StepApiResponse): Step {
  // Build complete screenshot URL
  // Backend returns: /api/runs/{run_id}/screenshots/{step}
  // We need: http://localhost:8080/api/runs/{run_id}/screenshots/{step}
  let screenshotUrl = step.screenshot_url || ''
  if (screenshotUrl && !screenshotUrl.startsWith('http')) {
    // If it's a relative URL starting with /api, use API_BASE_FOR_IMAGES
    if (screenshotUrl.startsWith('/api/')) {
      screenshotUrl = `${API_BASE_FOR_IMAGES}${screenshotUrl}`
    } else {
      // Otherwise prepend the full API base
      screenshotUrl = `${API_BASE_FOR_IMAGES}/api${screenshotUrl}`
    }
  }

  return {
    index: step.step_index,
    action: step.action,
    reasoning: step.reasoning,
    screenshot: screenshotUrl,
    status: step.status as 'success' | 'failed',
    error: step.error || undefined,
    duration_ms: step.duration_ms,
  }
}

function transformAssertionResult(result: AssertionResultApiResponse): AssertionResult {
  return {
    id: result.id,
    run_id: result.run_id,
    assertion_id: result.assertion_id,
    status: result.status as 'pass' | 'fail',
    message: result.message,
    actual_value: result.actual_value,
    created_at: result.created_at,
  }
}

export interface ReportDetailResponse extends Report {
  steps: Step[]
  assertion_results?: AssertionResult[]
  ui_assertion_results?: AssertionResult[]
  api_assertion_results?: AssertionResult[]
  pass_rate?: string
  api_pass_rate?: string
}

export async function listReports(params?: ReportsListParams): Promise<{ reports: Report[]; total: number; page: number; page_size: number }> {
  const searchParams = new URLSearchParams()
  if (params?.status) searchParams.set('status', params.status)
  if (params?.date) searchParams.set('date', params.date)
  if (params?.page) searchParams.set('page', params.page.toString())
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString())

  const query = searchParams.toString()
  const response = await apiClient<ReportsListApiResponse>(`/reports${query ? `?${query}` : ''}`)

  return {
    reports: response.reports.map(transformReport),
    total: response.total,
    page: response.page,
    page_size: response.page_size,
  }
}

export async function getReport(reportId: string): Promise<ReportDetailResponse> {
  const response = await apiClient<ReportDetailApiResponse>(`/reports/${reportId}`)

  return {
    ...transformReport(response),
    steps: response.steps.map(transformStep),
    assertion_results: (response.assertion_results || []).map(transformAssertionResult),
    ui_assertion_results: (response.ui_assertion_results || []).map(transformAssertionResult),
    api_assertion_results: (response.api_assertion_results || []).map(transformAssertionResult),
    pass_rate: response.pass_rate,
    api_pass_rate: response.api_pass_rate,
  }
}
